File: Algorithms/1. Sorting/quick-sort/quickSortHoare.py
""" Hoare partition scheme Quick Sort"""

# partition the array
def partition(array, left, right):
  pivot = array[(left+right)/2]; # choose rightmost element for comparison

  # swap elements < pivot with current element in a loop
  while left <= right:
    while array[left] < pivot: 
      left+=1; # index of left element should be on the right 
    while array[right] > pivot: 
      right-=1; # index of right element should be on the left 

    # swap elements and move left and right indices
    if left <= right:
      array[left], array[right] = array[right], array[left]
      left+=1
      right-=1
  return left;

# divide array into halves
def quick_sort(array, left, right):
  if left < right:
    pivot = partition(array, left, right)
    quick_sort(array, left, pivot - 1)
    quick_sort(array, pivot + 1, right)

# testing the algorithm
print("\nHoare partition scheme------")
print("Quick sort O(nlogn) algorithm:\n")

# array before sorting
array = [9, 8, 7, 6, 5, 4, 3, 2, 1]
print("Array before sorting")
print(array)

# test the quick sort algorithm
print("Array after sorting")
quick_sort(array, 0, len(array)-1)
print(array)

File: Algorithms/1. Sorting/quick-sort/quickSortHoare.cpp
// Hoare partition scheme Quick Sort
#include <iostream>
#include <vector>
using namespace std;

// swap function
void swap(int* num1, int* num2) {
  int temp = *num1;
  *num1 = *num2;
  *num2 = temp;
}

// Lomuto partition scheme
int partition(vector<int>& array, int left, int right) {
  int pivot = array[(left+right)/2]; // choose rightmost element for comparison

  // swap elements < pivot with current element in a loop
  while (left <= right) {
    while (array[left] < pivot) left++; // left element should be on the right 
    while (array[right] > pivot) right--; // right element should be on the left 

    // swap elements and move left and right indices
    if (left <= right)
      swap(array[left++], array[right--]);
  }
  return left;
}

// divide the array and sort it
void quickSort(vector<int>& array, int left, int right) {
  if (left < right) {
    int pivot = partition(array, left, right);
    quickSort(array, left, pivot-1);
    quickSort(array, pivot+1, right);
  }
}

int main() {
  cout << "Lomuto partition scheme------\n";
  cout << "Quick sort O(nlogn) algorithm\n";
  vector<int> array{5,4,3,2,1};
  cout << "Before sorting: ";
  for (auto num: array)
    cout << num << " ";
  cout << endl;

  // run quicksort algorithm
  quickSort(array, 0, array.size()-1);
  cout << "After sorting: ";
  for (auto num: array)
    cout << num << " ";
  cout << endl;
}